rustc_tools_util/src/lib.rs

```diff
@@ -104,10 +104,11 @@ impl std::fmt::Debug for VersionInfo {
 #[must_use]
 pub fn get_commit_hash() -> Option<String> {
     let output = std::process::Command::new("git")
-        .args(["rev-parse", "--short", "HEAD"])
+        .args(["rev-parse", "HEAD"])
```

Comment:
`rustc` uses `--short=9` in `bootstrap`. But that would result in _at least_ 9 characters. IIUC it will only use more characters, if 9 chars would not lead to a unique SHA. I think this is desirable.

https://github.com/rust-lang/rust/blob/2b78d920964e1d70927bcd208529bda0e11120d0/src/bootstrap/src/utils/channel.rs#L69-L71

Reply:
I am planning to remove `--short` from there too since, as you allude to, "at least 9" is still a source of nondeterminism that is unjustified.

"At least 9" is not actually better than "exactly 10" in practice. In fact, "at least 9" gives rise to 8&times; more collisions than "exactly 10", so it's worse from that perspective. For every new commit added to the repo, the probability that it makes a previously-unambiguous 9-digit hash become ambiguous is 16&times; higher than the probability that it collides with any previous 10-digit hash (offset by a factor of 2 because the second situation results in 2 ambiguous hashes, not just one).

In general, `--short=N` is not meaningful to use when the repository history is going to be mutated between when the short hash is generated until when a user makes use of the short hash.

Reply:
> In general, `--short=N` is not meaningful to use when the repository history is going to be mutated between when the short hash is generated until when a user makes use of the short hash.

That is a really good argument. I didn't think this through in that "direction".

I just ran [this script](https://stackoverflow.com/a/37403152) I found on SO. The Clippy repo today needs 9 digits to produce a unique hash for all existing commits:

```
$ git rev-list --abbrev=4 --abbrev-commit --all | ( while read -r line; do echo ${#line}; done; ) | sort -n | uniq -c
   1735 4
  17526 5
   3084 6
    203 7
     19 8
      1 9
```

The Rust repo itself already requires 10 digits:

```
$ git rev-list --abbrev=4 --abbrev-commit --all | ( while read -r line; do echo ${#line}; done; ) | sort -n | uniq -c
  20978 5
 206615 6
  36783 7
   2438 8
    151 9
      8 10
```

So to be on the save side and future proof, lets go with 10.

Reply:
It doesn't mean that 8 commits share the same 9 characters, just that there are 8 pairs of refs that have the same first 8 characters. The refs don't have to be 2 commits with the same first 9 characters. `git` also takes `blob`s and `tree`s into account:

```
$ git rev-list --abbrev=4 --abbrev-commit --all | ( while read -r line; do if [ ${#line} -eq 10 ]; then echo ${line::-1}; fi; done; ) | xargs -I {} git switch -d {}
error: short object ID f2dc18d0a is ambiguous
hint: The candidates are:
hint:   f2dc18d0a118 commit 2023-12-01 - update addr docs
hint:   f2dc18d0a09c tree
fatal: invalid reference: f2dc18d0a
error: short object ID e76b3f3b5 is ambiguous
hint: The candidates are:
hint:   e76b3f3b5b1b commit 2022-04-09 - Rename `unsigned_offset_from` to `sub_ptr`
hint:   e76b3f3b5d22 blob
fatal: invalid reference: e76b3f3b5
error: short object ID 2c4337a70 is ambiguous
hint: The candidates are:
hint:   2c4337a70a4d commit 2021-02-10 - Comments :3
hint:   2c4337a7029e blob
fatal: invalid reference: 2c4337a70
error: short object ID 01c625617 is ambiguous
hint: The candidates are:
hint:   01c6256178fb commit 2020-07-02 - Change debuginfo to default to 1 if `debug = true` is set
hint:   01c6256173f0 blob
fatal: invalid reference: 01c625617
error: short object ID 576294237 is ambiguous
hint: The candidates are:
hint:   576294237b10 commit 2018-01-29 - fix typos
hint:   576294237f84 tree
fatal: invalid reference: 576294237
error: short object ID 0b2e6f808 is ambiguous
hint: The candidates are:
hint:   0b2e6f808769 commit 2014-09-07 - Relicense shootout-reverse-complement.rs to the shootout license.
hint:   0b2e6f808075 tree
fatal: invalid reference: 0b2e6f808
error: short object ID 2d7b690b2 is ambiguous
hint: The candidates are:
hint:   2d7b690b2eaa commit 2013-12-09 - Register new snapshots
hint:   2d7b690b2ab9 tree
fatal: invalid reference: 2d7b690b2
error: short object ID 7ee90cc7b is ambiguous
hint: The candidates are:
hint:   7ee90cc7be7b commit 2012-04-24 - syntax: Rename is_word to is_keyword, etc.
hint:   7ee90cc7b29f tree
fatal: invalid reference: 7ee90cc7b
```

In fact, no commits share the first 9 chars. I don't know if this is for the better or worse.

How did you come up with the 50% number? I couldn't follow the logic.